src/openstack_cloud/openstack_manager.py

```diff
@@ -399,3 +445,305 @@ def create_instance(
+            addresses: The IP addresses to try SSH into.
+        """
+        try:
+            if not self._ssh_health_check(instance_name, addresses):
+                raise RunnerStartError(
+                    (
+                        "Unable to find running process of runner application on openstack runner "
+                        f"{instance_name}"
+                    )
+                )
+        except TimeoutError as err:
+            raise RunnerStartError(
+                f"Unable to connect to openstack runner {instance_name}"
+            ) from err
+
+    def _create_runner(self, conn: OpenstackConnection) -> None:
+        """Create a runner on OpenStack cloud.
+
+        Args:
+            conn: The connection object to access OpenStack cloud.
+        """
+        environment = jinja2.Environment(
+            loader=jinja2.FileSystemLoader("templates"), autoescape=True
+        )
+
+        env_contents = _generate_runner_env(
+            templates_env=environment,
+            proxies=self._config.charm_state.proxy_config,
+            dockerhub_mirror=self._config.dockerhub_mirror,
+            ssh_debug_connections=self._config.charm_state.ssh_debug_connections,
+        )
+        instance_config = create_instance_config(
+            self.app_name, self.unit_num, IMAGE_NAME, self._config.path, self._github
+        )
+        cloud_userdata = _generate_cloud_init_userdata(
+            templates_env=environment, instance_config=instance_config, runner_env=env_contents
+        )
+
+        self._ensure_security_group(conn)
+        self._setup_runner_keypair(conn, instance_config.name)
+        try:
+            instance = conn.create_server(
+                name=instance_config.name,
+                image=IMAGE_NAME,
+                key_name=instance_config.name,
+                flavor=self._config.flavor,
+                network=self._config.network,
+                security_groups=[SECURITY_GROUP_NAME],
+                userdata=cloud_userdata,
+                timeout=180,
+                wait=True,
+            )
+        except openstack.exceptions.ResourceTimeout as err:
+            logger.exception("Timeout creating OpenStack runner %s", instance_config.name)
+            try:
+                conn.delete_server(name_or_id=instance_config.name, wait=True)
+            except openstack.exceptions.SDKException:
+                logger.critical(
+                    "Cleanup of creation failure runner %s has failed", instance_config.name
+                )
+                # Reconcile will attempt to cleanup again prior to spawning new runners.
+            raise RunnerCreateError(
+                f"Timeout creating OpenStack runner {instance_config.name}"
+            ) from err
+
+        addresses = [address["addr"] for address in instance.addresses[self._config.network]]
+        self._wait_until_runner_process_running(instance.name, addresses)
```

Comment:
There is a race condition here where if the job is picked up and exits before this check, this will keep waiting.